javascript/ql/src/ApiGraphs.qll

```diff
@@ -0,0 +1,749 @@
+        rhs = exp.getSourceNode(prop)
+        or
+        exists(Variable v |
+          exp.exportsAs(v, prop) and
+          rhs = v.getAnAssignedExpr().flow()
+        )
+      )
+    }
+
+    private DataFlow::SourceNode trackUseNode(DataFlow::SourceNode nd, DataFlow::TypeTracker t) {
+      t.start() and
+      use(_, nd) and
+      result = nd
+      or
+      exists(DataFlow::TypeTracker t2 | result = trackUseNode(nd, t2).track(t2, t))
+    }
+
+    /**
+     * Gets a node that is inter-procedurally reachable from `nd`, which is a use of some feature.
+     */
+    cached
+    DataFlow::SourceNode trackUseNode(DataFlow::SourceNode nd) {
+      result = trackUseNode(nd, DataFlow::TypeTracker::end())
+    }
+
+    private DataFlow::SourceNode trackDefNode(DataFlow::Node nd, DataFlow::TypeBackTracker t) {
+      t.start() and
+      def(_, nd) and
+      result = nd.getALocalSource()
+      or
+      exists(DataFlow::TypeBackTracker t2 | result = trackDefNode(nd, t2).backtrack(t2, t))
+    }
+
+    /**
+     * Gets a node that inter-procedurally flows into `nd`, which is a definition of some feature.
+     */
+    cached
+    DataFlow::SourceNode trackDefNode(DataFlow::Node nd) {
+      result = trackDefNode(nd, DataFlow::TypeBackTracker::end())
+    }
+
+    private DataFlow::SourceNode getANodeWithType(TypeName tn) {
+      exists(string moduleName, string typeName |
+        tn.hasQualifiedName(moduleName, typeName) and
+        result.hasUnderlyingType(moduleName, typeName)
+      )
+    }
+
+    /**
+     * Holds if there is an edge from `pred` to `succ` in the API graph that is labeled with `lbl`.
+     */
+    cached
+    predicate edge(TFeature pred, string lbl, TFeature succ) {
+      exists(string m |
+        pred = MkRoot() and
+        lbl = Label::mod(m) and
+        succ = MkModule(m)
+      )
+      or
+      exists(string m |
+        pred = MkModule(m) and
+        lbl = Label::member("exports")
+      |
+        succ = MkModuleExport(m)
+        or
+        succ = MkModuleImport(m)
+      )
+      or
+      exists(DataFlow::SourceNode ref |
+        use(pred, lbl, ref) and
+        succ = MkUse(ref)
+      )
+      or
+      exists(DataFlow::Node rhs |
+        def(pred, lbl, rhs) and
+        succ = MkDef(rhs)
+      )
+      or
+      exists(DataFlow::Node def |
+        def(pred, def) and
+        lbl = Label::instance() and
+        succ = MkClassInstance(trackDefNode(def))
+      )
+      or
+      exists(CanonicalName cn |
+        pred = MkRoot() and
+        lbl = Label::mod(cn.getExternalModuleName())
+      |
+        succ = MkCanonicalNameUse(cn) or
+        succ = MkCanonicalNameDef(cn)
+      )
+      or
+      exists(CanonicalName cn1, CanonicalName cn2 |
+        cn2 = cn1.getAChild() and
+        lbl = Label::member(cn2.getName())
+      |
+        (pred = MkCanonicalNameDef(cn1) or pred = MkCanonicalNameUse(cn1)) and
+        (succ = MkCanonicalNameDef(cn2) or succ = MkCanonicalNameUse(cn2))
```

Comment:
Perhaps set literals would make this a little nicer

Reply:
Doesn't immediately work due to type inference failure. Do you want me to define auxiliary predicates wrapping `MkCanonicalName{Def,Use}` so I can use them here?

Reply:
Ah of course, let's just leave it like this then.